src/librustc_typeck/check/mod.rs

```diff
@@ -4210,6 +4214,141 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         ty
     }
 
+    /// Given a `NodeId`, return the `FnDecl` of the method it is enclosed by and wether it is
+    /// `fn main` if it is a method, `None` otherwise.
+    pub fn get_fn_decl(&self, blk_id: ast::NodeId) -> Option<(hir::FnDecl, bool)> {
+        // Get enclosing Fn, if it is a function or a trait method
+        let fn_id = self.tcx.hir.get_parent(blk_id);
+        let parent = self.tcx.hir.get(fn_id);
+
+        if let Node::NodeItem(&hir::Item {
+            name, node: hir::ItemFn(ref decl, ..), ..
+        }) = parent {
+            decl.clone().and_then(|decl| {
+                Some((decl, name == Symbol::intern("main")))
```

Comment:
It seemed good enough at the time, given that the failure mode is *not* suggesting a return type when it could, but point taken. Is there a way to figure out wether the current method is *the* main method during `typeck`?